lib/fileset.nix

```
@@ -0,0 +1,993 @@
+    else if pathType root != "directory" then
+      throw "lib.fileset.toSource: Expected attribute `root` \"${toString root}\" to be a path pointing to a directory, but it's pointing to a file instead."
+    else if ! hasPrefix root fileset._base then
+      throw "lib.fileset.toSource: Expected attribute `fileset` to not be influenceable by any paths outside `root`, but `lib.fileset.getInfluenceBase fileset` \"${toString fileset._base}\" is outside `root`."
+    else
+      cleanSourceWith {
+        name = "source";
+        src = root;
+        filter = pathString: _: inSet (fromSubpath "./${pathString}");
+      };
+
+  /*
+  Create a file set from a filtered local source as produced by the `lib.sources` functions.
+  This does not import anything into the store.
+
+  Type:
+    fromSource :: SourceLike -> FileSet
+
+  Example:
+    fromSource (lib.sources.cleanSource ./.)
+  */
+  fromSource = source:
+    if ! source._isLibCleanSourceWith or false || ! source ? origSrc || ! source ? filter then
+      throw "lib.fileset.fromSource: Expected the argument to be a value produced from `lib.sources`, but got a ${typeOf source} instead."
+    else if ! isPath source.origSrc then
+      throw "lib.fileset.fromSource: Expected the argument to be source-like value of a local path."
+    else
+      _fromSource source.origSrc source.filter;
+
+  /*
+  Coerce a value to a file set:
+
+  - If the value is a file set already, return it directly
+
+  - If the value is a path pointing to a file, return a file set with that single file
+
+  - If the value is a path pointing to a directory, return a file set with all files contained in that directory
+
+  This function is mostly not needed because all functions in `lib.fileset` will implicitly apply it for arguments that are expected to be a file set.
+
+  Type:
+    coerce :: Any -> FileSet
+  */
+  coerce = value: _coerce "coerce" "argument" value;
+
+
+  /*
+  Create a file set containing all files contained in a path (see `coerce`), or no files if the path doesn't exist.
+
+  This is useful when you want to include a file only if it actually exists.
+
+  Type:
+    optional :: Path -> FileSet
+  */
+  optional = path:
+    if ! isPath path then
+      throw "lib.fileset.optional: Expected argument to be a path, but got a ${typeOf path}."
+    else if pathExists path then
+      _singleton path
+    else
+      _create path null;
+
+  /*
+  Return the common ancestor directory of all file set operations used to construct this file set, meaning that nothing outside the this directory can influence the set of files included.
+
+  Type:
+    getInfluenceBase :: FileSet -> Path
+
+  Example:
+    getInfluenceBase ./Makefile
+    => ./.
+
+    getInfluenceBase ./src
+    => ./src
+
+    getInfluenceBase (fileFilter (file: false) ./.)
+    => ./.
+
+    getInfluenceBase (union ./Makefile ../default.nix)
+    => ../.
+  */
+  getInfluenceBase = maybeFileset:
```

```suggestion
  commonPathPrefix = maybeFileset:
```

```suggestion
  commonRoot = maybeFileset:
```

I know what you mean from the descriptions, but "influence base" is not a helpful name.